lang/macro/src/codegen/ink_test.rs

```diff
@@ -0,0 +1,88 @@
+// Copyright 2018-2019 Parity Technologies (UK) Ltd.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use derive_more::From;
+use proc_macro2::TokenStream as TokenStream2;
+use quote::quote;
+
+pub use crate::{
+    codegen::{
+        cross_calling::CrossCalling,
+        dispatch::Dispatch,
+        env_types::EnvTypes,
+        events::{
+            EventHelpers,
+            EventStructs,
+        },
+        metadata::GenerateMetadata,
+        storage::Storage,
+        GenerateCode,
+        GenerateCodeUsing,
+    },
+    ir::InkTest,
+};
+
+/// Generates code for the `[ink::test]` macro.
+#[derive(From)]
+pub struct InkTestModule<'a> {
+    /// The test function to generate code for.
+    ink_test: &'a InkTest,
+}
+
+impl GenerateCode for InkTestModule<'_> {
+    /// Generates the code for `#[ink:test]`.
+    fn generate_code(&self) -> TokenStream2 {
+        let item_fn = &self.ink_test.item_fn;
+        let attrs = &item_fn.attrs;
+        let sig = &item_fn.sig;
+        let fn_name = &sig.ident;
+        let fn_return_type = &sig.output;
+        let fn_block = &item_fn.block;
+        let _vis = &item_fn.vis;
+        let _fn_args = &sig.inputs;
+        match fn_return_type {
+            syn::ReturnType::Default => {
+                quote! {
+                    #( #attrs )*
+                    #[test]
+                    fn #fn_name() {
+                        env::test::run_test::<env::DefaultEnvTypes, _>(|_| {
+                            #fn_block
+                            Ok(())
+                        })
+                        .expect(&format!("{}: the off-chain testing environment returned an error", stringify!(#fn_name)));
+                    }
+                }
+            }
+            syn::ReturnType::Type(_rarrow, _type) => {
+                quote! {
+                    #( #attrs )*
+                    #[test]
+                    fn #fn_name() -> env::Result<()> {
+                        env::test::run_test::<env::DefaultEnvTypes, _>(|_| {
+                            #fn_block
```

Comment:
```suggestion
                            { #fn_block }
```

Reply:
Also I think in this case where the user wants to return a `Result` we do not need to append the `Ok(())` in order to make their lifes easier because it won't.

Reply:
Can you elaborate how it won't make their life easier? If we omit the `Ok(())` here they would have to add it manually in every test.

Reply:
it is more transparent as to what they would write if their test function was returning a `Result`. after all we should only generate barely enough magic as to make this still appear transparent enough for users to derive a sense of it. I would definitely be baffled by having an implicit return of `Ok((())` that I do not see myself.